In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [2]:
credit_card_data = pd.read_csv('/content/creditcard.csv')

Time = Number of seconds elapsed between this transaction and the first transaction in the dataset

V1-V28 = result of a PCA Dimensionality reduction to protect user identities and sensitive features


In [3]:
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [4]:
credit_card_data['Class'].value_counts()

0.0    17836
1.0       81
Name: Class, dtype: int64

In [5]:
legit_data = credit_card_data[credit_card_data.Class == 0]
fraud_data = credit_card_data[credit_card_data.Class == 1]

In [7]:
print(legit_data.shape)
print(fraud_data.shape)

(17836, 31)
(81, 31)


Count: There are 17,836 data points in the dataset.

Mean: The average value of the data points is 67.365.

Standard Deviation (std): The data is quite spread out from the mean, with a standard deviation of 188.75. This suggests high variability in the values.

Minimum (min): The smallest value in the dataset is 0.

25th Percentile (25%): 25% of the values are less than or equal to 5.49.

50th Percentile (50%) or Median: The middle value in the dataset is 15.95. This means half of the values are below 15.95, and half are above it.

75th Percentile (75%): 75% of the values are less than or equal to 56.2325.

Maximum (max): The largest value in the dataset is 7712.43.

In [8]:
legit_data.Amount.describe()

count    17836.000000
mean        67.365025
std        188.754429
min          0.000000
25%          5.490000
50%         15.950000
75%         56.232500
max       7712.430000
Name: Amount, dtype: float64

In [9]:
fraud_data.Amount.describe()

count      81.000000
mean       98.105926
std       267.464067
min         0.000000
25%         1.000000
50%         1.000000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [10]:
credit_card_data.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                    
0.0    13891.131980 -0.203851  0.228560   0.840631  0.263581 -0.117213   
1.0    16833.074074 -9.315066  6.779465 -13.052870  6.451318 -6.557541   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.112151 -0.110725 -0.008454  0.752065  ...  0.032646 -0.054930   
1.0   -2.602613 -9.012856  4.730365 -3.087651  ...  0.756424  0.549752   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.144997 -0.036871  0.015838  0.118904  0.035813  0.010352  0.006670   
1.0   -0.423495 -0.398104 -0.277836  0.355688  0.177616  0.940693  0.090007   

          Amount  
Class             
0.0    67.365025  
1.0    98.105926  

[2 rows x 30 columns]

In [11]:
legit_sample = legit_data.sample(n=81)

In [14]:
new_dataset = pd.concat([legit_sample, fraud_data], axis=0)

In [15]:
new_dataset['Class'].value_counts()

0.0    81
1.0    81
Name: Class, dtype: int64

In [17]:
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

Stratify is a technique used to improve the quality and representativeness of your training and testing datasets. It involves dividing the data into subgroups (also called strata) based on one or more important features. Then, randomly select a proportionate number of samples from each subgroup to create your training and testing sets. This ensures that the sets reflect the overall distribution of the original data even if there are imbalances.

In [19]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [20]:
print(X.shape, X_train.shape, X_test.shape)

(162, 30) (129, 30) (33, 30)


In [30]:
model = LogisticRegression(solver='liblinear')

In [ ]:
model.fit(X_train, Y_train)

In [32]:
X_pred = model.predict(X_train)
training = accuracy_score(X_pred, Y_train)

In [33]:
print('The accuracy on Training data : ', training)

The accuracy on Training data :  0.9922480620155039


In [34]:
X_test_pred = model.predict(X_test)
testing = accuracy_score(X_test_pred, Y_test)

In [35]:
print('The accuracy on Testing Data : ', testing)

The accuracy on Testing Data :  0.8787878787878788
